report_docx/report/report_docx.py

```diff
@@ -0,0 +1,306 @@
+            'pdf': 'report.pdf',
+            'docx': 'report.docx'
+        }
+
+        self._delete_temp_folder(tmp_folder_name)
+        self._create_temp_folder(tmp_folder_name)
+
+        self._generate_reports(
+            cr, uid, context, tmp_folder_name, data,
+            output_type, output_report)
+
+        report = self._get_convert_file(
+            tmp_folder_name, output_report[output_type])
+
+        self._delete_temp_folder(tmp_folder_name)
+
+        return (report, output_type)
+
+    def generate_docx_data(self, cr, uid, ids, context):
+        """
+            Override this method to pass your own data to the engine.
+            The return value of this module should be a list with
+            report data.
+        """
+        return [{}]
+
+    def _generate_reports(
+        self, cr, uid, context, tmp_folder_name,
+        datas, output_type, output_report
+    ):
+        if "pdf" == output_type:
+            self._generate_pdf_reports(
+                cr, uid, context, tmp_folder_name, datas,
+                output_type, output_report)
+            return
+
+        self._generate_doc_reports(
+            cr, uid, context, tmp_folder_name, datas,
+            output_type, output_report)
+
+    def _generate_pdf_reports(
+        self, cr, uid, context, tmp_folder_name,
+        datas, output_type, output_report
+    ):
+        count = 0
+        for data in datas:
+            self._convert_single_report(
+                cr, uid, context, tmp_folder_name, count, data, output_type)
+            count = count + 1
+
+        self._combine_pdf_files(
+            tmp_folder_name, output_report[output_type])
+
+    def _generate_doc_reports(
+        self, cr, uid, context, tmp_folder_name,
+        datas, output_type, output_report
+    ):
+        if len(datas) > 1:
+            raise Exception(_(
+                "Multiple docx files export doesn't support. " +
+                "Please export single docx file."))
+        temp_report = self._convert_single_report(
+            cr, uid, context, tmp_folder_name, 0, datas[0], output_type)
+
+        cmd = "mv %s %s" % (
+            tmp_folder_name + temp_report,
+            tmp_folder_name + output_report[output_type])
+
+        os.system(cmd)
+
+    def _combine_pdf_files(self, tmp_folder_name, output_report):
+        output_path = tmp_folder_name + output_report
+        output_temp_path = tmp_folder_name + 'temp.pdf'
+
+        cmd = "gs -q -dNOPAUSE -sDEVICE=pdfwrite -sOUTPUTFILE=%s \
+            -dBATCH %s*water*.pdf" % (output_temp_path, tmp_folder_name)
+        os.system(cmd)
+
+        # remove the last empty page
+        input_stream = PdfFileReader(file(output_temp_path, 'rb'))
+        output_stream = PdfFileWriter()
+
+        pagenum = input_stream.getNumPages()
+        for i in range(pagenum - 1):
+            page = input_stream.getPage(i)
+            output_stream.addPage(page)
+
+        out_stream = file(output_path, 'wb')
```

Comment:
Same note regarding this file handler & needing to be in a context manager. The `handler.close` is then implicitly called no matter what happens in terms of errors.